### PR TITLE
Issue 18149 lettuce cache transport

### LINKED ISSUE #18149: Redis / Lettuce CacheTransport and Provider
dotCMS needs to standardize on an external caching layer. We have hedged our bets between Hazelcast and Redis and ultimately chose to include Hazelcast as it provided most functionality out of the box, including a pub/sub channel that we could use to invalidate caches between dotCMS instances.

Redis 5.0 provides a data type called Streams which is well suited to act as a lossless pub/sub queue, needed for our cache transport layer in dotCMS.  You can read more about it here:  https://redis.io/topics/streams-intro.

The redis java library `lettuce` https://github.com/lettuce-io/lettuce-core allows us to and subscribe to these streams in redis and use them to provide a log of cache invalidations.  The "log" like nature of Redis streams will allow a dotCMS instance to receive invalidations asynchronously, that were sent when it is not in a state to do so, for example when a node is being cycled or during a period of high GC activity.

We need:

1. `RedisPubSubImpl`
2. `RedisCacheTransportImpl` that uses the `RedisPubSubImpl`
3. `RedisCacheProvider` that uses Lettuce and allows us to use Redis as a CacheProvider

### PR DESCRIPTION
**Pull request by wezell**
This is the cache + transport + transactional improvements for redis

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/cache/lettuce/Message.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
+package com.dotcms.cache.lettuce;
```
Comment (wezell) [inline dotCMS/src/main/java/com/dotcms/cache/lettuce/Message.java:1]: We should not need this class as Redis should use the new PubSub framework

File: dotCMS/src/integration-test/java/com/dotcms/MainSuite.java [hunk 2]
```diff
@@ -456,7 +458,8 @@
         BundlePublisherTest.class,
         CategoryFactoryTest.class,
         Task210805DropUserProxyTableTest.class,
-        Task210816DeInodeRelationshipTest.class,
+        DotObjectCodecTest.class,
```
Comment (nollymar) [inline dotCMS/src/integration-test/java/com/dotcms/MainSuite.java:461]: Is correct the place where these tests are being added? why not at the end? 

Besides, I see two new tests that aren't included here: 
- LettuceCacheTest.java
- RedisPubSubImplTest.java

File: dotCMS/src/integration-test/java/com/dotcms/cache/lettuce/LettuceCacheTest.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    public static void startup() throws Exception {
+        cache = new RedisCache();
+    }
+    
+    
+    /**
+     * all key values that are put into redis are prefixed by a random key.
+     * When a flushall is called, this key is cycled, which basically invalidates
+     * all cached entries.  The prefix key is stored in redis itself so multiple 
+     * servers in the c
```
Comment (nollymar) [inline dotCMS/src/integration-test/java/com/dotcms/cache/lettuce/LettuceCacheTest.java:26]: This comment seems to be incomplete

File: dotCMS/src/integration-test/java/com/dotcms/cache/lettuce/LettuceCacheTest.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    @Test
+    public void test_cache_key_generation() {
+        
+        final String prefix  = cache.loadPrefix();
+        assert(prefix!=null && prefix.length()>3);
+        String group = "group";
+        String key = "key";
+        String cacheKey = cache.cacheKey(group, key);
+        assert (cacheKey!=null);
+        assert (cacheKey.startsWith(prefix));
```
Comment (nollymar) [inline dotCMS/src/integration-test/java/com/dotcms/cache/lettuce/LettuceCacheTest.java:55]: These asserts: `assert (cacheKey.startsWith(prefix));` and `assert (cacheKey.startsWith(prefix+ "." + group + "."));` are unnecessary if on the next line you write: `assert (cacheKey.equals(prefix+ "." + group+ "." + key));`

File: dotCMS/src/integration-test/java/com/dotcms/cache/lettuce/LettuceCacheTest.java [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        assert (cacheKey.startsWith(prefix+ "." + group + "."));
+        assert (cacheKey.equals(prefix+ "." + group+ "." + key));
+
+        String cacheKey2 = cache.cacheKey(group, null);
+        assert (cacheKey2!=null);
+        assert (cacheKey2.startsWith(prefix));
```
Comment (nollymar) [inline dotCMS/src/integration-test/java/com/dotcms/cache/lettuce/LettuceCacheTest.java:61]: This assert is unnecessary too: `assert (cacheKey2.startsWith(prefix));`

File: dotCMS/src/main/java/com/dotcms/cache/lettuce/DotObjectCodec.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    @Override
+    public ByteBuffer encodeKey(final K key) {
+
+        return null != key? charset.encode(key.toString()): null;
+    }
+
+    @Override
+    public ByteBuffer encodeValue(final V value) {
+
+        try (ByteArrayOutputStream baos = new ByteArrayOutputStream();
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotcms/cache/lettuce/DotObjectCodec.java:50]: `baos` and `output` can be declared final

File: dotCMS/src/main/java/com/dotcms/cache/lettuce/DotPubSubListener.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        if (this.channelSet.contains(channel)) {
+
+            this.messageConsumer.accept(message);
+        }
+    }
+
+    @Override
+    public void message(K pattern, K channel, V message) {
+
+        if (this.channelSet.contains(channel)) {
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotcms/cache/lettuce/DotPubSubListener.java:39]: where is `pattern` used here? If not used, it should call `message(channel, message)`

**Review by nollymar — CHANGES_REQUESTED** [review]

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/system/cache/CacheResource.java [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+
+/**
+ * Cache Resource for diff provides
+ * @author jsanca
+ */
+@Path("/v1/caches")
+public class CacheResource {
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/system/cache/CacheResource.java:37]: This resource should have postman tests

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/system/redis/RedisResource.java [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+import javax.ws.rs.core.Response;
+import java.util.Set;
+import java.util.concurrent.Future;
+
+/**
+ * Redis Resource
+ * @author jsanca
+ */
+@Path("/v1/redis")
+public class RedisResource {
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/system/redis/RedisResource.java:35]: This resource should have postman tests

File: dotCMS/src/main/java/com/dotcms/rest/config/DotRestApplication.java [hunk 10]
```diff
@@ -170,6 +172,7 @@
 			.add(VersionableResource.class)
 			.add(PermissionResource.class)
 			.add(ContentResource.class)
+			.add(CacheResource.class)
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotcms/rest/config/DotRestApplication.java:175]: Shouldn't `RedisResource` be added here?

File: dotCMS/src/main/java/com/dotmarketing/portlets/containers/model/FileAssetContainer.java [hunk 11]
```diff
@@ -135,11 +143,11 @@ public void setPath(final String path) {
     }
 
     private Versionable toContentlet() {
-
-        contentlet.setIdentifier(this.identifier);
-        contentlet.setInode(this.inode);
-        contentlet.setLanguageId(this.languageId);
-        return contentlet;
+        Contentlet testContentlet =  new Contentlet();
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotmarketing/portlets/containers/model/FileAssetContainer.java:146]: Rename testContentlet

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/system/redis/RedisResource.java (hunk 9 above)
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/rest/api/v1/system/redis/RedisResource.java:35]: will mark visible for testing, bc we do not want to share this one

File: dotCMS/src/main/java/com/dotcms/rest/config/DotRestApplication.java (hunk 10 above)
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/rest/config/DotRestApplication.java:175]: nope, b/c it was just for own testing

**Review by nollymar — APPROVED** [review]

**Review by erickgonzalez — APPROVED** [review]

**Pull request merged**